### PR TITLE
Catch unhandled build exceptions

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

flutter assemble catches unhandled build exceptions, but the buildAssemble bundle implementation does. Example:

```
	at _Directory._fillWithDirectoryListing	(directory_patch.dart:37 )
at _Directory.listSync	(directory_impl.dart:249 )
at ForwardingDirectory.listSync	(forwarding_directory.dart:45 )
at _AssetDirectoryCache.variantsFor	(asset.dart:517 )
at _parseAssetFromFile	(asset.dart:659 )
at _parseAssets	(asset.dart:583 )
at _ManifestAssetBundle.build	(asset.dart:151 )
at <asynchronous gap>	(async )
at AssetOutputBehavior.outputs	(assets.dart:77 )
at SourceVisitor.visitBehavior	(source.dart:171 )
at _SourceBehavior.accept	(source.dart:247 )
at Target._resolveConfiguration	(build_system.dart:230 )
at Target.resolveOutputs	(build_system.dart:188 )
at Target._toNode	(build_system.dart:132 )
at BuildSystem.build	(build_system.dart:415 )
at _AsyncAwaitCompleter.start	(async_patch.dart:43 )
at BuildSystem.build	(build_system.dart:400 )
at buildWithAssemble	(bundle.dart:125 )
at _AsyncAwaitCompleter.start	(async_patch.dart:43 )
at buildWithAssemble	(bundle.dart:99 )
at BundleBuilder.build	(bundle.dart:75 )
at _AsyncAwaitCompleter.start	(async_patch.dart:43 )
at BundleBuilder.build	(bundle.dart:52 )
at IOSSimulator._sideloadUpdatedAssetsForInstalledApplicationBundle	(simulators.dart:453 )
at IOSSimulator._setupUpdatedApplicationBundle	(simulators.dart:420 )
at _AsyncAwaitCompleter.start	(async_patch.dart:43 )
at IOSSimulator._setupUpdatedApplicationBundle	(simulators.dart:419 )
at IOSSimulator.startApp	(simulators.dart:352 )
at _AsyncAwaitCompleter.start	(async_patch.dart:43 )
at IOSSimulator.startApp	(simulators.dart:339 )
at FlutterDevice.runHot	(resident_runner.dart:393 )
at _asyncThenWrapperHelper.<anonymous closure>	(async_patch.dart:71 )
at _rootRunUnary	(zone.dart:1132 )
at _CustomZone.runUnary	(zone.dart:1029 )
at _FutureListener.handleValue	(future_impl.dart:137 )
at Future._propagateToListeners.handleValueCallback	(future_impl.dart:678 )
at Future._propagateToListeners	(future_impl.dart:707 )
at Future._completeWithValue	(future_impl.dart:522 )
at _AsyncAwaitCompleter.complete	(async_patch.dart:30 )
at _completeOnAsyncReturn	(async_patch.dart:288 )
at BuildableIOSApp.fromProject	(application_package.dart )
at _asyncThenWrapperHelper.<anonymous closure>	(async_patch.dart:71 )
at _rootRunUnary	(zone.dart:1132 )
at _CustomZone.runUnary	(zone.dart:1029 )
at _FutureListener.handleValue	(future_impl.dart:137 )
at Future._propagateToListeners.handleValueCallback	(future_impl.dart:678 )
at Future._propagateToListeners	(future_impl.dart:707 )
at Future._completeWithValue	(future_impl.dart:522 )
at Future._asyncComplete.<anonymous closure>	(future_impl.dart:552 )
at _rootRun	(zone.dart:1124 )
at _CustomZone.run	(zone.dart:1021 )
at _CustomZone.runGuarded	(zone.dart:923 )
at _CustomZone.bindCallbackGuarded.<anonymous closure>	(zone.dart:963 )
at _microtaskLoop	(schedule_microtask.dart:41 )
at _startMicrotaskLoop	(schedule_microtask.dart:50 )
at _runPendingImmediateCallback	(isolate_patch.dart:116 )
at _RawReceivePortImpl._handleMessage	(isolate_patch.dart:173 )
```

### CODE REVIEW

File: packages/flutter_tools/lib/src/bundle.dart [hunk 1]
```diff
@@ -122,22 +122,26 @@ Future<void> buildWithAssemble({
...
+      final Depfile depfile = Depfile(result.inputFiles, result.outputFiles);
+      final File outputDepfile = fs.file(depfilePath);
+      if (!outputDepfile.parent.existsSync()) {
+        outputDepfile.parent.createSync(recursive: true);
+      }
+      depfile.writeToFile(outputDepfile);
     }
-    depfile.writeToFile(outputDepfile);
+  } catch (err) {
+    throwToolExit('Failed to build bundle.$err');
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/bundle.dart]: In other `throwToolExit` places we have a `: ` before the error.
`Failed to build bundle: $err`

File: packages/flutter_tools/lib/src/bundle.dart [hunk 2]
```diff
@@ -122,22 +122,26 @@ Future<void> buildWithAssemble({
...
-  if (!result.success) {
-    for (ExceptionMeasurement measurement in result.exceptions.values) {
-      printError(measurement.exception.toString());
-      printError(measurement.stackTrace.toString());
+    if (!result.success) {
+      for (ExceptionMeasurement measurement in result.exceptions.values) {
+        printError(measurement.exception.toString());
+        printError(measurement.stackTrace.toString());
+      }
+      throwToolExit('Failed to build bundle.');
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/bundle.dart]: Does this not get caught?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/bundle.dart]: This only catches errors which happen during the build target's execution, and not permission errors hit during the build system itself

File: packages/flutter_tools/lib/src/bundle.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/bundle.dart]: Made this the same as the toolExit above

File: packages/flutter_tools/lib/src/bundle.dart (hunk 2 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/bundle.dart]: I guess I'm asking if throwing a tool exit here will get caught and wrapped in a second tool exit?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/bundle.dart]: oh...hmmm. I"m being silly

**Comment by jonahwilliams** [comment]
Closing because I forgot how exceptions work

**Pull request closed without merging**